doc/developer/chips/10.rst

```diff
@@ -531,24 +531,76 @@ For more details on when the copy initializer would be called, please refer to
+ * Class variables are deinitialized when `delete` is called on them.
+
+To specify actions that occur upon deinitialization, user can specify a `proc
 deinit()` method. This method must have zero arguments and cannot be called
 directly. If a type does not specify a `proc deinit()` method, the compiler
 will generate one.
+It will be equivalent to a user-written `deinit` with an empty body.
+
+In addition to the actions specified by the `deinit` method, whether
+compiler-generated or user-defined, in the process of deinitialization:
 
-No matter whether the `deinit` method is compiler generated or user-defined,
-the compiler will ensure that in the process of deinitialization, all fields of
-record type are deinitialized. Additionally, the compiler will ensure that
-parent `deinit` methods are called in the case of inherited classes.
+ * parent `deinit` methods are called in the case of inherited classes, and
+ * all fields of record type are deinitialized.
 
 The expected use-case for the `deinit` method is to call `delete` on
 fields of class type, when appropriate. It is the responsibility of the
 class designer to `delete` fields when appropriate. The compiler will not
 automatically add `delete` calls for fields of class type.
+Conversely, the compiler adds `deinit` calls on fields of record type.
+These calls cannot be added by the user.
+
+Ensuring type safety
+********************
+
+We would like to ensure that a field cannot be accessed after it has been
+deinitialized. In the deinitialization process, the field `F` of a record
+or class `RC` could potentially be accessed in the following cases:
+
+ (a) From `RC.deinit()` or functions it invokes.
+
+ (b) From `deinit()` in a child of `RC` (if `RC` is a class),
+     via direct reference or a function call not involving
+     dynamic dispatch on `this`.
+
+ (c) From `deinit` in a parent of `RC` by invoking a method on `this`
+     that dynamically dispatches to `RC` or RC's child.
+
+ (d) In the case of dependencies between fields within `RC`, when
+     deinitializing a field that depends on `F`.
+     Recall that field initialization order prevents fields of
+     a child class to depend on fields of a parent class.
+
+As a simple example of field dependency in (d), in the following code
+the field `A` depends on the field `D`:
+
+.. code-block:: chapel
+
+    record R {
+      const n = 10;
+      const D = {1..n};
+      var A: [D] real;
+      ...
+    }
+
+Our current proposal handles these cases as follows:
+
+ (a) `RC.deinit()` is executed before `F` is deinitialized.
+
+ (b) `deinit()` in children of `RC`, if any, are invoked
+     before `RC.deinit()` and therefore before `F` is deinitialized.
+
+ (c) Dynamic dispatch on `this` is disallowed in `deinit` methods.
+
+ (d) Dependent fields within `RC` are deinitialized
+     in reverse declaration order.
+
+This approach is detailed below. It is a work in progress.
```

Comment:
I agree with Michael that this section is too detailed.  Do we not already mention all of these requirements?  Why not put the short rationale for why each limitation is necessary closer to the introduction of the limitation?